mfa/push/src/main/kotlin/com/pingidentity/mfa/push/PingAMPushHandler.kt

```diff
@@ -0,0 +1,370 @@
+                }
+            }
+            
+            // Add all extracted data to the result map
+            result[KEY_MESSAGE_ID] = messageId
+            if (mechanismUid != null) {
+                result[KEY_CREDENTIAL_ID] = mechanismUid
+            }
+            if (challenge != null) {
+                result[KEY_CHALLENGE] = challenge
+            }
+            result[KEY_TTL] = ttl
+            if (loadBalancer != null) {
+                result[KEY_AMLB_COOKIE] = loadBalancer
+            }
+            if (customMessage != null) {
+                result[KEY_MESSAGE_TEXT] = customMessage
+            }
+            if (customPayload != null) {
+                result[KEY_CUSTOM_PAYLOAD] = customPayload
+            }
+            if (numbersChallenge != null) {
+                result[KEY_NUMBERS_CHALLENGE] = numbersChallenge
+            }
+            if (contextInfo != null) {
+                result[KEY_CONTEXT_INFO] = contextInfo
+            }
+            if (pushType != null) {
+                result[KEY_PUSH_TYPE] = pushType
+            }
+            if (userId != null) {
+                result[KEY_USER_ID] = userId
+            }
+            if (timeIntervalString != null) {
+                try {
+                    val timeInterval = timeIntervalString.toLong()
+                    result[KEY_TIME_INTERVAL] = timeInterval
+                } catch (e: NumberFormatException) {
+                    logger.w("Time interval was not a valid number: $timeIntervalString. Error: ${e.message}")
+                }
+            }
+            
+            // Store the original data for reference
+            result[KEY_RAW_JWT] = message
+
+            // Additional data includes everything from claimsMap not already extracted
+            val additionalData = claimsMap.filterKeys {
+                it !in listOf(
+                    JWT_MECHANISM_UID, JWT_CHALLENGE, JWT_TTL, JWT_TIME_INTERVAL,
+                    JWT_NOTIFICATION_MESSAGE, JWT_PUSH_TYPE, JWT_CUSTOM_PAYLOAD,
+                    JWT_NUMBERS_CHALLENGE, JWT_CONTEXT_INFO, JWT_AM_LOAD_BALANCER_COOKIE,
+                    JWT_USER_ID
+                )
+            }
+            result[KEY_ADDITIONAL_DATA] = additionalData
+            
+            logger.d("Successfully parsed PingAM push notification with messageId: $messageId")
+        } catch (e: Exception) {
+            logger.e("Failed to parse PingAM push notification: ${e.message}", e)
+        }
+        
+        return result
+    }
+    
+    /**
+     * Send an approval response for a notification.
+     *
+     * @param credential The credential to use for the response.
+     * @param notification The notification to approve.
+     * @return True if the approval was sent successfully, false otherwise.
+     */
+    override suspend fun sendApproval(
+        credential: PushCredential,
+        notification: PushNotification,
+        params: Map<String, Any>
```

Comment:
For the params, can we use data class instead?

Reply:
No, same as above, we don't know the payload from other providers. Also, in push notification world, usually a map or json string is given by the providers.